### ISSUE TITLE
Infinite Dimensional Hyperparameters

### ISSUE THREAD

**Issue by azev77** [alan-turing-institute/MLJ.jl#477]
Some models allow infinite dimensional hyper-parameters. 
**Eg 1**: [XGBoost](https://juliacomputing.com/blog/2020/02/24/ad-xgboost.html) allows custom loss functions. (As does @xiaodaigh's [JLBoost.jl](https://github.com/xiaodaigh/JLBoost.jl))
**Eg 2**: @joshday's [SparseRegression.jl](https://github.com/joshday/SparseRegression.jl) allows custom loss & custom penalty
```Julia
s = SModel(x, y, L2DistLoss(), L2Penalty())
```
**Eg 3**: @rakeshvar's    [AnyBoost.jl](https://discourse.julialang.org/t/anyboost-jl/25769) allows custom Loss/Activation/Constraint 

An oft advertised feature of "doing ML in Julia" is how easy it is to customize traditional models: 
([Julia computing](https://juliacomputing.com/blog/2020/02/24/ad-xgboost.html)) & ([Discourse](https://discourse.julialang.org/t/custom-xgboost-loss-function-w-zygote-julia-computing-blog-post/35811))

1. Is it possible to make it easy to identify which models in MLJ allow infinite dim HP (custom Loss/Penalty/CEF) 
For example:
`models(x -> x.is_supervised && x.is_pure_julia && x.custom_penalty)`
lists all supervised models written in pure julia which allow custom penalties...

2. It would be awesome to automate tuning of infinite dimensional HP
The [Julia computing](https://juliacomputing.com/blog/2020/02/24/ad-xgboost.html) example adds weight parameter `w` to the logistic loss:
![image](https://user-images.githubusercontent.com/7883904/77484023-e7b97300-6dff-11ea-8dc6-a777a83069c5.png)
It would be cool to make it easy tune `w` over a grid in MLJ... 

3. At some point, it would be magical to include a model in MLJ called `GenericLearner` w/ 3 infinite dimensional inputs: (L(y,ŷ), f(x), P(f)) where:
L(y, ŷ; θ_L): the **Loss function**, in the above example `θ_L = w`
ŷ = f(x; θ_f): the **Model**, in SparseRegression.jl it is always `f(x; θ_f)=  θ_f*x`
P(f();θ_P): the **Penalty**, L1/MCP what suits your fancy as long as the opt problem is `nice`
Then the user chooses her fav algorithm to solve: 
ŷ = arginf Q():= L(y, ŷ; θ_L) + P(f)
ŷ = arginf Q():= L(y, f(x; θ_f); θ_L) + P(θ_f; θ_P)
I realize this may be insanely hard to solve unless the user makes judicious choices of (L, f, P). 

Some joke that this is "[All of ML in one expression](https://fxdiebold.blogspot.com/2017/01/all-of-machine-learning-in-one.html)". It's not but it is a powerful abstraction.

@rakeshvar put it [nicely](https://discourse.julialang.org/t/new-machine-learning-package-julia-implementation-of-xgboost/8912/13?u=albert_zevelev) when discussing [JuML.jl](https://github.com/Statfactory/JuML.jl): 

>  One of the main points of Gradient Boosting (and extensions like XGBoost) is its application/generalization to any loss (not just linear regression or logistic regression). This generalization is very easily achieved via Multiple Dispatch in julia. In that context, seeing that the package works only for logistic was surprising. Additionally, the code seems to be hard-coded around Logit loss.
It would be great to abstract away the XGBoost algorithm from the direct dependence on loss, and instead, the algorithm can get the gradient (and hessian) from the loss Type (this is where multiple dispatch is so handy). This makes adding new losses very easy.

**Comment by azev77** [comment]
Also a generic learner is a great pedagogical tool b/c it compactly unifies many different ML models.
I'd love to co-author a tutorial that gives a non-black box intro to ML. 

First consider constant models: ŷ = f(x; θ_f) = θ_f
(1.a) Mean: (L(y,ŷ) = (y-ŷ)^2, f(x)= θ_f, P(f)=0)
(1.b) Median: (L(y,ŷ) = |y-ŷ|, f(x)= θ_f, P(f)=0)
(1.c) Quantile: (L(y,ŷ; θ_L) = Check(), f(x)= θ_f, P(f)=0).  Where θ_L is the `quantile`.

Next consider linear (in the params) models: ŷ = f(x; θ_f) = θ_f*x
(2.a) OLS: (L(y,ŷ) = (y-ŷ)^2, f(x)= θ_f*x, P(f)=0)
(2.b) Lasso: (L(y,ŷ) = (y-ŷ)^2, f(x)= θ_f*x, P(f)=L1())
(2.c) Ridge: (L(y,ŷ) = (y-ŷ)^2, f(x)= θ_f*x, P(f)=L2())
(2.d) LAD: (L(y,ŷ) = |y-ŷ|, f(x)= θ_f*x, P(f)=0)
(2.e) Quantile Reg: (L(y,ŷ) = Check(), f(x)= θ_f*x, P(f)=0)
